spec/helpers/checkout_helper_spec.rb

```diff
@@ -15,12 +15,27 @@
     helper.validated_input("test", "foo", type: :email)
   end
 
-  describe "displaying the tax total for an order" do
-    let(:order) { double(:order, total_tax: 123.45, currency: 'AUD') }
+  describe "#display_checkout_tax_total" do
+    subject(:display_checkout_tax_total) { helper.display_checkout_tax_total(order) }
+
+    let(:order) { instance_double(Spree::Order, total_tax: 123.45, currency: 'AUD') }
+    let(:service) { instance_double(VoucherAdjustmentsService, voucher_included_tax: ) }
```

Comment:
I would have set up real data and used the real VoucherAdjustmentsService. What if  the method `VoucherAdjustmentsService#voucher_included_tax` go renamed one day, or returned a different value? This test would still pass.

But maybe that's just opinion :)

If it's because the test data is complicated to set up, then maybe we need to create a factory for it?

Reply:
`instance_double` verifies that the methods are present. https://www.simplybusiness.co.uk/about-us/tech/2020/05/rspec-instance-doubles/

But in general I agree that mocking is to be avoided if possible.

Reply:
Thanks, I didn't know that!